uamqp/aio/_cbs_async.py

```diff
@@ -0,0 +1,227 @@
+        self._token_status_description = status_description
+
+        if execute_operation_result == ManagementExecuteOperationResult.OK:
+            self.auth_state = CbsAuthState.Ok
+        elif execute_operation_result == ManagementExecuteOperationResult.ERROR:
+            self.auth_state = CbsAuthState.Error
+            # put-token-message sending failure, rejected
+            self._token_status_code = 0
+            self._token_status_description = "Auth message has been rejected."
+        elif execute_operation_result == ManagementExecuteOperationResult.FAILED_BAD_STATUS:
+            self.auth_state = CbsAuthState.Error
+
+    def _update_status(self):
+        # TODO: this is duplication the sync version, considering mixin or inheritance
+        if self.state == CbsAuthState.Ok or self.state == CbsAuthState.RefreshRequired:
+            is_expired, is_refresh_required = self._check_expiration_and_refresh_status()
+            if is_expired:
+                self.state = CbsAuthState.Expired
+            elif is_refresh_required:
+                self.state = CbsAuthState.RefreshRequired
+        elif self.state == CbsAuthState.InProgress:
+            put_timeout = self._check_put_timeout_status()
+            if put_timeout:
+                self.state = CbsAuthState.Timeout
+
+    def _check_expiration_and_refresh_status(self):
+        # TODO: this is duplication the sync version, considering mixin or inheritance
+        seconds_since_epoc = int(utc_now().timestamp())
+        is_expired = seconds_since_epoc >= self._expires_on
+        is_refresh_required = (self._expires_on - seconds_since_epoc) <= self._refresh_window
+        return is_expired, is_refresh_required
+
+    def _check_put_timeout_status(self):
+        # TODO: this is duplication the sync version, considering mixin or inheritance
+        if self._auth_timeout > 0:
+            return (int(utc_now().timestamp()) - self._token_put_time) >= self._auth_timeout
+        else:
+            return False
+
+    def _cbs_link_ready(self):
+        # TODO: this is duplication the sync version, considering mixin or inheritance
+        if self.state == CbsState.OPEN:
+            return True
+        if self.state != CbsState.OPEN:
+            return False
+        if self.state in (CbsState.CLOSED, CbsState.ERROR):
+            # TODO: raise proper error type
+            raise AuthenticationException("CBS authentication link is in broken status, please recreate the cbs link.")
+
+    async def open(self):
+        self.state = CbsState.OPENING
+        await self._mgmt_link.open()
+
+    async def close(self):
+        await self._mgmt_link.close()
+        self.state = CbsState.CLOSED
+
+    async def update_token(self):
+        self.auth_state = CbsAuthState.InProgress
+        access_token = await self._auth.get_token()
+        self._expires_on = access_token.expires_on
+        self._expires_in = self._expires_on - int(utc_now().timestamp())
```

Comment:
Do we actually need/use _both_ expires_on and expires_in?

Reply:
we only need _expires_on, _expires_in is used to calculate refresh window.

I have updated the code to make _expires_in a local variable